### ISSUE TITLE
55 - Union to Intersection

### ISSUE THREAD

**Issue by githubxiaowen** [type-challenges/type-challenges#122]
```ts
type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
```
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#type-inference-in-conditional-types

**Comment by Sociosarbis** [comment]
```typescript
type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
type I = UnionToIntersection<'foo' | 42 | true> // never

type A = 'foo' | 42 | true

type B = A extends any ? (arg: A) => any : never // (arg: A) => any 

type C = B extends ((arg: infer I) => void) ? I : never //  'foo' | 42 | true

type D = 'foo' & 2 & true // never
```
Hi, @githubxiaowen ,I try your solution in my editor.It works when all statements combined as a type template.But When I separate those statements into three statements in order to figure out what mechanism your solution is based on, it didn't work as expect.

Could you give me more hints to understand the underlying magic of the solution. (I have read the link you attached above.)

BTW, do you know the reason why the three separated statement doesn't work the same as the template type ?

**Comment by jo32** [comment]
> ```ts
> type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
> type I = UnionToIntersection<'foo' | 42 | true> // never
> 
> type A = 'foo' | 42 | true
> 
> type B = A extends any ? (arg: A) => any : never // (arg: A) => any 
> 
> type C = B extends ((arg: infer I) => void) ? I : never //  'foo' | 42 | true
> 
> type D = 'foo' & 2 & true // never
> ```
> 
> Hi, @githubxiaowen ,I try your solution in my editor.It works when all statements combined as a type template.But When I separate those statements into three statements in order to figure out what mechanism your solution is based on, it didn't work as expect.
> 
> Could you give me more hints to understand the underlying magic of the solution. (I have read the link you attached above.)
> 
> BTW, do you know the reason why the three separated statement doesn't work the same as the template type ?

@Sociosarbis It is because `'foo' & 42 & true` is logically `never` -- you can never find a type who is `42` and `true` at the same time.

the magic of this solution is given a type like `(arg: U) => any | (arg: V) => any`, if you want to find a argument(using keyword infer) satisfying U and V, the type of it must be U & V;

**Comment by swnb** [comment]
amazing!

**Comment by vogler** [comment]
If the intersection for a field is never, you can't provide a value for the object anymore.
So when using `UnionToIntersection` on objects out of your control the below `StripNever` may be useful.

~~~typescript
type t = UnionToIntersection<{a: number, b: number} | {a: number, b: string}>;
const t1: t = {a: 1}; //  Property 'b' is missing in type '{ a: number; }' but required in type '{ a: number; b: number; }'
const t2: t = {a: 1, b: 1}; // Type 'number' is not assignable to type 'never'.

type NonNeverKeys<T> = { [K in keyof T]: T[K] extends never ? never : K }[keyof T];
type StripNever<T> = T extends object ? { [K in NonNeverKeys<T>]: StripNever<T[K]> } : T;

const t3: StripNever<t> = {a: 1}; // ok
~~~

**Comment by Jackie1210** [comment]
![image](https://user-images.githubusercontent.com/14737221/125602227-54d8ad0b-3bc2-4b39-8f2d-6c1017f57c21.png)

**Comment by wine-fall** [comment]
The solution is amazing, however, i think write as below may be understand more easily
```
type UnionToIntersection<U, T = U> =  (U extends T ? (arg: U) => any : never) extends (arg: infer I) => void ? I : never;
```
And what does the `U extends U` mean? (when `U` is an Union Type) you can check link: https://github.com/type-challenges/type-challenges/issues/614

**Comment by Alexsey** [comment]
Can somebody please explain why almost equivalent code is not working?
```ts
type UnionToIntersection<U> = U extends U ? (arg: U) => any extends (arg: infer I) => any ? I : never : never
```
It looks the same, but I have no explanations for the result, that is
```ts
((x: true) => unknown) | ((x: "foo") => unknown) | ((x: 42) => unknown)
```

**Comment by kstratis** [comment]
> 



> > ```ts
> > type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
> > type I = UnionToIntersection<'foo' | 42 | true> // never
> > 
> > type A = 'foo' | 42 | true
> > 
> > type B = A extends any ? (arg: A) => any : never // (arg: A) => any 
> > 
> > type C = B extends ((arg: infer I) => void) ? I : never //  'foo' | 42 | true
> > 
> > type D = 'foo' & 2 & true // never
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hi, @githubxiaowen ,I try your solution in my editor.It works when all statements combined as a type template.But When I separate those statements into three statements in order to figure out what mechanism your solution is based on, it didn't work as expect.
> > Could you give me more hints to understand the underlying magic of the solution. (I have read the link you attached above.)
> > BTW, do you know the reason why the three separated statement doesn't work the same as the template type ?
> 
> @Sociosarbis It is because `'foo' & 42 & true` is logically `never` -- you can never find a type which is `42` and `true` at the same time.
> 
> the magic of this solution is given a type like `(arg: U) => any | (arg: V) => any`, if you want to find an argument(using keyword infer) satisfying U and V, the type of it must be U & V;

You're right but I don't think that you really answered what @Sociosarbis actually asked: 
`type C` should have been `'foo' & 42 & true` (which is `never`) instead of `'foo' | 42 | true`. 

Putting this thing together returns the proper intersection but why separate statement don't work as expected here? Why putting one-by-one returns a union instead of an intersection? 

cc @jo32 

Thanks!

**Comment by kstratis** [comment]
> ```ts
> type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
> type I = UnionToIntersection<'foo' | 42 | true> // never
> 
> type A = 'foo' | 42 | true
> 
> type B = A extends any ? (arg: A) => any : never // (arg: A) => any 
> 
> type C = B extends ((arg: infer I) => void) ? I : never //  'foo' | 42 | true
> 
> type D = 'foo' & 2 & true // never
> ```
> 
> Hi, @githubxiaowen ,I try your solution in my editor.It works when all statements combined as a type template.But When I separate those statements into three statements in order to figure out what mechanism your solution is based on, it didn't work as expect.
> 
> Could you give me more hints to understand the underlying magic of the solution. (I have read the link you attached above.)
> 
> BTW, do you know the reason why the three separated statement doesn't work the same as the template type ?

I was wondering the exact same thing, so I turned to SO. 
Turns out your break-down is missing generics:
[https://stackoverflow.com/a/74003461/178728](https://stackoverflow.com/a/74003461/178728)

**Comment by Sociosarbis** [comment]
> > ```ts
> > type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
> > type I = UnionToIntersection<'foo' | 42 | true> // never
> > 
> > type A = 'foo' | 42 | true
> > 
> > type B = A extends any ? (arg: A) => any : never // (arg: A) => any 
> > 
> > type C = B extends ((arg: infer I) => void) ? I : never //  'foo' | 42 | true
> > 
> > type D = 'foo' & 2 & true // never
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Hi, @githubxiaowen ,I try your solution in my editor.It works when all statements combined as a type template.But When I separate those statements into three statements in order to figure out what mechanism your solution is based on, it didn't work as expect.
> > Could you give me more hints to understand the underlying magic of the solution. (I have read the link you attached above.)
> > BTW, do you know the reason why the three separated statement doesn't work the same as the template type ?
> 
> I was wondering the exact same thing, so I turned to SO. Turns out your break-down is missing generics: https://stackoverflow.com/a/74003461/178728

an answer with clear explanations. thx @kstratis

**Comment by quantum-rose** [comment]
I got never [ts playground](https://www.typescriptlang.org/play/?ts=5.5.2#code/C4TwDgpgBAqgdgSwPZwCpIJJ2BATgZwgGNhk4AeGAPigF4oAKGKCADxzgBN8oBDOEFAD8jXrgDmALlgBKOjX6DpcCADc8cth26MGYqVARwAZnigY5tGqqQJOckRijK1eAFChI5urEQp0WDgExKQo5ADkxkhI4VAAPlAALABM8VDAuACuEDQA9LksrJAkEJzpSFAARtCR0bEAZEmpjRnZQA)

**Comment by skeleton1231** [comment]
type UnionToIntersection<U> = 
  (U extends any ? (x: U) => void : never) extends (x: infer R) => void 
    ? R 
    : never;

**Comment by skeleton1231** [comment]
> ```ts
> ```ts
> type UnionToIntersection<U> = (U extends any ? (arg: U) => any : never) extends ((arg: infer I) => void) ? I : never
> ```
> 
> 
>     
>       
>     
> 
>       
>     
> 
>     
>   
> ```

  Expect<Equal<UnionToIntersection<never>, never>>,
how about this if single type also regard as union type?

**Comment by goddnsgit** [comment]
我看到了这句话,然后明白了.
多个函数类型的联合 (A => any) | (B => any)，在匹配 (infer I) => any 时，I 必须同时满足所有函数的参数类型，即 I 必须是 A & B。
